### ISSUE TITLE
Async wrappers can throw

### ISSUE THREAD

**Issue by dev7355608** [ruipin/fvtt-lib-wrapper#7]
```js
class Foo {
    async bar() {
        return 0;
    }
}

libWrapper.register("some-module", "Foo.prototype.bar", async function (wrapped) {
    await new Promise(resolve => setTimeout(resolve, 1000));
    const a = await wrapped();
    return a + 1;
});

const foo = new Foo();

await foo.bar();
```

```
Uncaught libWrapper: This wrapper function is no longer valid, and must not be called. This is most likely caused by an issue in the 'Foo.prototype.bar' wrapper registered by module 'some-module'.
```

It's probably necessary to add a new suffix to support asynchronous functions, because we can't know whether `wrapped` is called asynchronously or the function simply returns a promise but `wrapped` is called synchronously.

```js
libWrapper.register("some-module", "Foo.prototype.bar#async", ...)
```

**Comment by dev7355608** [comment]
Or we assume that, if the function returns a promise, `wrapped` is called asynchronously (i.e. called in that promise) as well, which is probably almost always the case. And we add the suffix `#sync` instead to indicate the opposite behavior. This way we would be backwards compatible.

**Comment by ruipin** [comment]
Thanks for the report. You are right, this is actually related to the limitations that used to exist (lifted in 0.0.7) regarding multiple chaining.

I did not think of the situation where people wouldn't awit for something other than the `wrapped` function, so never planned for this.

The problem is that as soon as you return from the wrapper, the `wrapped` method gets invalidated. This is to prevent breakage if e.g. the wrapper order changes.

I think it might be possible to lift the restrictions regarding calling wrapper functions asynchronously. There is already a mechanism to detect when a wrapper would become invalid because of e.g. a wrapper order change, I could make it so that only those invalidate the wrapper.

However, fully preventing this kind of invalidation is difficult. I can easily foresee an interaction between two modules, where one awaits for some event, and meanwhile the second registers a new wrapper to the same method, causing the previous wrapper to throw once it resolves.

I don't see an easy way to avoid this, though. I'll make sure to document this behaviour, and will revisit it later.

**Comment by dev7355608** [comment]
This doesn't work as well in 1.0.7.0, I noticed.

```js
libWrapper.register("midi-qol", "Foo.prototype.bar", async function (wrapped) {
    const a = await wrapped();
    const b = await wrapped();
    return a + b + 1;
});
```

Is it guaranteed that the first awaited function is called before the function returns? That's why it works if the first await is `wrapped`?

**Comment by ruipin** [comment]
Yeah, ends up something more or less equivalent `return wrapped(). then(/* rest of code here */)`. 

The first method is called immediately, the expectation being that it returns a promise.

Only the code after the await will be doem asynchronously.

**Comment by ruipin** [comment]
With [v1.0.8.0](https://github.com/ruipin/fvtt-lib-wrapper/releases/tag/v1.0.8.0), I've removed the hardcoded limitation preventing asynchronous chaining of wrappers. Now, they will only be invalidated if the internal data structures are changed.

I've also updated the documentation to discuss this use case and the limitations [here](https://github.com/ruipin/fvtt-lib-wrapper/blob/5b6110d6728226283af348fb2cbbf6cb61be9e31/README.md#registering-or-unregistering-a-wrapper-invalidates-any-pending-wrapper-chains-for-a-given-method).

I'll keep this issue open so that I can come back and revisit improving this remaining limitation once I have more time.

**Comment by dev7355608** [comment]
```js
const p1 = wrapped();
const p2 = wrapped();
const a = await p1;
const b = await p2;
```

This would solve my problem actually if I could guarantee the order of `p1` and `p2` (`p1` needs to be executed before `p2`). But that's not possible, I think.

**Comment by ruipin** [comment]
You can guarantee the order of execution by the order of `await`. The Promises will only be resolved once you `await` them.

The issue is if the side-effects pre-promise break the second await - for example, if the methods open a dialog, and the promise is waiting for the dialog to be closed. Such a solution would cause two dialogs to open simultaneously.

**Comment by dev7355608** [comment]
A promise can be executed before `await`. `p2` might be executed before `p1` even if we await `p1` and then `p2`. This example prints `hello!` without `await`:

```js
async function say(msg) {
    console.log(msg);
}
say("hello!");
```

**Comment by ruipin** [comment]
That method isn't executing a Promise though, or to be more exact, the Promise is empty.

That function is more or less equivalent to:
```javascript
function say(msg) {
    console.log(msg);
    return new Promise(resolve => resolve());
}
``` 

What you're seeing is that the side-effects of the method (i.e. code before returning a Promise, in this case the `console.log`) execute immediately, synchronously.

Notice how it returned a promise (using your code):
```javascript
let x = say("bla");
// 'bla'
// -> Promise { <state>: "fulfilled", <value>: undefined }
```

Now notice this:
```javascript
function say(msg) {
    console.log(msg);
    return new Promise(resolve => {});
}

let x = say("bla");
// 'bla'
// -> Promise { <state>: "pending" }
```

**Comment by dev7355608** [comment]
I see. I'm not too familiar with async/await.

The code in Midi QOL that causes this problem can probably removed anyway. But it's not the same problem that's causing all these Midi issues lately. I messed something else up when I was adding libWrapper compatibility. I uncovered this issue after I fixed the other problems.

Thanks for your quick response and effort to fix this problem!

**Comment by ruipin** [comment]
[v1.1.0.0](https://github.com/ruipin/fvtt-lib-wrapper/releases/tag/v1.1.0.0) (specifically commit https://github.com/ruipin/fvtt-lib-wrapper/commit/eeacc6710d5842f3a8aa6eaaf2b704c3e0ebcbf9) adds a bunch of async tests to the test suite and fixes an issue related to async 'WRAPPER'-type wrappers.

The mechanism to detect failure to chain the wrapper further would fire before the Promise was allowed to run, therefore unregister wrappers incorrectly.

**Comment by ruipin** [comment]
[v1.3.0.0](https://github.com/ruipin/fvtt-lib-wrapper/releases/tag/v1.3.0.0) refactors the test suite such that separate tests are not required to test the async code path. Now, every test will run twice - once synchronously, and a second time asynchronous.

This should make it much easier to touch this section of code in the future, and ensures the async code path is as well tested as the sync code path without duplicating work.

**Comment by ruipin** [comment]
Fixed in [v1.3.1.0](https://github.com/ruipin/fvtt-lib-wrapper/releases/tag/v1.3.1.0).

Starting with this version, modifying a wrapper (e.g. using libWrapper.register/unregister) will not trigger invalidation of any outstanding wrappers. Outstanding wrappers will continue execution *as if nothing had happened*, i.e. they will not see the new wrappers or wrapper order.
